### ISSUE TITLE
TLS Certificate Pinning

### ISSUE THREAD

**Issue by nitesharora91** [eclipse-paho/paho.mqtt.c#456]
Hi
In my case, I have a Mosquiito Broker running on a local machine, with dynamic IP. (Actually, machine can be connected to any WiFi as a part of system configuration by the user). 

It is not possible to specify the IP addresses as a part of Subject Alternative Name (SAN).

The client is an Android Device, which is supposed to be in the same LAN (WiFi) as the mosquitto machine. Android Client doesn't have any provision for hostname mapping in /etc/hosts.

The only way to connect to Broker is using the Local IP of Mosquiito Machine.

Since the certificates are generated with BrokerMachine's hostname as CN, whenever the MQTT Client tries to connect to the Broker using IP address of the BrokerMachine, SSL error of unmatching Hostname and CN is encountered.

Is there any way to forge the server Hostname  at client while making the connection using IP?
Actually, what is required is something similar to Curl's --resolve option.

**From Curl's manpage: Provide a  custom  address  for  a specific host and port pair. Using  this, you can make the curl requests(s) use a specified address and   prevent the otherwise normally resolved address to be used. Consider   it a sort of /etc/hosts alternative provided on the command line**

Is there any similar way this can be achieved with Paho MQTT client. Any analogous method to curl's--resolve? 
Or any other method that could help to resolve this issue.
I do not wish to use insecure connection even in LAN.

Regards
Nitesh

**Comment by icraggs** [comment]
1) So what you're suggesting is a connect option which contains a hostname which is purely used for the hostname check on the TLS certificate presented by the server?  (If I understand it correctly).  You can't do this today.

2) It doesn't sound very secure - you would be forcing the check against a hostname which isn't actually mapped to the address being connected to.

**Comment by nitesharora91** [comment]
Yes, Ian, I am exactly looking for something you mentioned in point 1.

If that is not possible, any available option or tweak for **server certificate pinning** that could be used?

**Comment by jumoog** [comment]
good reference:

https://www.owasp.org/index.php/Certificate_and_Public_Key_Pinning#OpenSSL

**Comment by DBL2017** [comment]
need an external callback for certificate verification.
